### ISSUE TITLE
Receive "django.core.exceptions.SynchronousOnlyOperation" when adding user permissions from Django db to JWT claims

### ISSUE THREAD

**Issue by itsjef** [supertokens/supertokens-python#294]
I followed the [Adding custom claims to the JWT during creation](https://supertokens.com/docs/session/common-customizations/sessions/with-jwt/custom-claims#adding-custom-claims-to-the-jwt-during-creation) tutorial to add user's object-level permissions from Django database to JWT claims. 

The titled exception appeared for every method I applied for defining/calling `get_all_permissions`, including:
1. defining it as an asynchronous function using `async def`.
2. defining a synchronous version of it, decorated with [`@sync_to_async`](https://docs.djangoproject.com/en/4.1/topics/async/#sync-to-async).
3. still the sync function but instead of decorating with `@sync_to_async`, I called in a separate thread as suggested in the original traceback (the traceback was a bit different in this case but still the same exception).

I ran out of idea to try here, so I kindly ask for support from Supertokens team. Could you suggest a way to make this work? Thanks!

---

## Code
```python
def session_override(original_implementation: RecipeInterface):
    og_create_new_session = original_implementation.create_new_session

    async def get_all_permissions(user_id: str):
        UserModel = get_user_model()

        try:
            user = await UserModel.objects.aget(pk=user_id)
            return user.get_all_permissions()
        except UserModel.DoesNotExist:
            return

    async def create_new_session(
        request: Any,
        user_id: str,
        access_token_payload: Union[None, Dict[str, Any]],
        session_data: Union[None, Dict[str, Any]],
        user_context: Dict[str, Any],
    ):
        if access_token_payload is None:
            access_token_payload = {}

        permissions = await get_all_permissions(user_id)
        access_token_payload["scope"] = " ".join(permissions) if permissions else ""

        return await og_create_new_session(
            request, user_id, access_token_payload, session_data, user_context
        )

    original_implementation.create_new_session = create_new_session
    return original_implementation


init(
    app_info=InputAppInfo(
        app_name=APP_NAME,
        api_domain=API_DOMAIN,
        api_base_path=API_BASE_PATH,
        website_domain=WEBSITE_DOMAIN,
        website_base_path=WEBSITE_BASE_PATH,
    ),
    supertokens_config=SupertokensConfig(
        connection_uri=AUTH_SERVICE_URL,
    ),
    framework="django",
    recipe_list=[
        session.init(override=session.InputOverrideConfig(functions=session_override)),
        emailpassword.init(),
    ],
    mode="wsgi",
)

```

## Traceback
```python
Internal Server Error: /api/auth/signin
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.10/site-packages/supertokens_python/framework/django/django_middleware.py", line 72, in __syncMiddleware
    result: Union[DjangoResponse, None] = async_to_sync(st.middleware)(
  File "/usr/local/lib/python3.10/site-packages/asgiref/sync.py", line 218, in __call__
    return call_result.result()
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 439, in result
    return self.__get_result()
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.10/site-packages/asgiref/sync.py", line 284, in main_wrap
    result = await self.awaitable(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/supertokens_python/supertokens.py", line 598, in middleware
    api_resp = await matched_recipe.handle_api_request(
  File "/usr/local/lib/python3.10/site-packages/supertokens_python/recipe/emailpassword/recipe.py", line 183, in handle_api_request
    return await handle_sign_in_api(self.api_implementation, api_options)
  File "/usr/local/lib/python3.10/site-packages/supertokens_python/recipe/emailpassword/api/signin.py", line 42, in handle_sign_in_api
    response = await api_implementation.sign_in_post(
  File "/usr/local/lib/python3.10/site-packages/supertokens_python/recipe/emailpassword/api/implementation.py", line 171, in sign_in_post
    session = await create_new_session(
  File "/usr/local/lib/python3.10/site-packages/supertokens_python/recipe/session/asyncio/__init__.py", line 71, in create_new_session
    return await SessionRecipe.get_instance().recipe_implementation.create_new_session(
  File "/app/app/settings/common.py", line 410, in create_new_session
    permissions = await get_all_permissions(user_id)
  File "/app/app/settings/common.py", line 380, in get_all_permissions
    user = await UserModel.objects.defer("groups", "user_permissions").aget(pk=user_id)
  File "/usr/local/lib/python3.10/site-packages/django/db/models/query.py", line 662, in aget
    return await sync_to_async(self.get)(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/asgiref/sync.py", line 435, in __call__
    ret = await asyncio.wait_for(future, timeout=None)
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 408, in wait_for
    return await fut
  File "/usr/local/lib/python3.10/site-packages/asgiref/current_thread_executor.py", line 22, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.10/site-packages/asgiref/sync.py", line 476, in thread_handler
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/django/db/models/query.py", line 646, in get
    num = len(clone)
  File "/usr/local/lib/python3.10/site-packages/django/db/models/query.py", line 376, in __len__
    self._fetch_all()
  File "/usr/local/lib/python3.10/site-packages/django/db/models/query.py", line 1866, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/usr/local/lib/python3.10/site-packages/django/db/models/query.py", line 87, in __iter__
    results = compiler.execute_sql(
  File "/usr/local/lib/python3.10/site-packages/django/db/models/sql/compiler.py", line 1396, in execute_sql
    cursor = self.connection.cursor()
  File "/usr/local/lib/python3.10/site-packages/django/utils/asyncio.py", line 24, in inner
    raise SynchronousOnlyOperation(message)
django.core.exceptions.SynchronousOnlyOperation: You cannot call this from an async context - use a thread or sync_to_async.
```

## Versions
- Python 3.11
- Django 4.1.2
- supertokens-python 0.11.12

**Comment by itsjef** [comment]
I tried defining `create_new_session` as a synchronous function and wrapped with with `sync_to_async` before reassigning to `original_implementation.create_new_session` but it raised

```python
RuntimeError: This event loop is already running
```

**Comment by KShivendu** [comment]
That's surprising. I tried something similar and it worked for me.

```python
def override_session_functions(oi: SessionRecipeImplementation):
    oi_create_new_session = oi.create_new_session

    async def get_all_permissions(user_id):
        from .models import UserLastSeen
        try:
            author = await Authors.objects.aget(user_id=user_id)
            return ["author"]
        except Exception as e:
            return []

    async def create_new_session(request, user_id, access_token_payload, session_data, user_context):
        if access_token_payload is None:
            access_token_payload = {}

        permissions = await get_all_permissions(user_id)
        access_token_payload["scope"] = " ".join(permissions)
        print(permissions)

        new_session = await oi_create_new_session(request, user_id, access_token_payload, session_data, user_context)
        return new_session

    oi.create_new_session = create_new_session
    return oi
```

Can you recheck your django version and tell me how are you running it?
I got `django.core.exceptions.SynchronousOnlyOperation: You cannot call this from an async context - use a thread or sync_to_async` only when I tried to call `UserLastSeen.objects.get(user_id=user_id)` without await inside `async get_all_permissions()`

Also, in case you want to use, sync_to_async. You can try:

```python
from asgiref.sync import sync_to_async

def get_all_permissions(user_id: str):
        UserModel = get_user_model()

        try:
            user = UserModel.objects.get(pk=user_id) # .aget() won't work anymore 
            return user.get_all_permissions()
        except UserModel.DoesNotExist:
            return [] # Note that returning [] makes it cleaner :)

async def create_new_session():
   ...
   permissions = await sync_to_async(get_all_permissions)(user_id)
   access_token_payload['scope'] = " ".join(permissions)
   ...
```

Let me know if it helps :D

**Comment by KShivendu** [comment]
Ohh wait. You're using:
 
```python
user = await UserModel.objects.defer("groups", "user_permissions").aget(pk=user_id)
```

I doubt if `.defer()` isn't aligned with django's new async ORM features. If it runs a sync ORM query, I'd expect django to throw `django.core.exceptions.SynchronousOnlyOperation`

**Comment by itsjef** [comment]
Other query versions I have tried are:
```python
user = UserModel.objects.all().get(pk=user_id)

user = UserModel.objects.all().filter(pk=user_id).get()

user = UserModel.objects.all().filter(pk=user_id).first()

user = list(UserModel.objects.all().filter(pk=user_id))[0]  # as suggested in: https://stackoverflow.com/a/68760464
```
(notes that `all()` and `filter(...)` can be interchanged)

**Comment by itsjef** [comment]
Please help re-open the issue since it is still unresolved. I don't have the permission to do so.

**Comment by KShivendu** [comment]
> except Exception therefore it would catch SynchronousOnlyOperation and, as a result, your get_all_permissions would always returns []

Ahh good Point. Actually, I wanted to be able to catch all the errors so I used Exception. But the thing is, I ran the debugger as well and my query was giving me the output as I expected. Plus, the end result wasn't [] but ["foo"].

Can you try replicating this issue in a cloud environment like Repl.it or gitpod?

**Comment by KShivendu** [comment]
> except Exception therefore it would catch SynchronousOnlyOperation and, as a result, your get_all_permissions would always returns []

Ahh good Point. Actually, I wanted to be able to catch all the errors so I used Exception. But the thing is, I ran the debugger as well and my query was giving me the output as I expected. Plus, the end result wasn't [] but ["foo"].

Can you try replicating this issue in a cloud environment like Repl.it or gitpod?

**Comment by itsjef** [comment]
I have created an example here: https://github.com/itsjef/django-supertokens-example

Anyway, you are right @KShivendu. The problem is not with `supertokens-python`, but `gunicorn`. I have been using `gevent` worker class for Gunicorn, which doesn't play well with asyncio:

```shell
$ gunicorn -k gevent -b '0.0.0.0:8000' app.wsgi:application
```

The problem wouldn't happen if I used Django's built-in development server (`python manage.py runserver`).

In case of Gunicorn, I'm gonna have to revert to the default `sync` worker, I am unsure if that would affect performance but that's another story.
